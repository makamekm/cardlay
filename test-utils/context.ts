import { Context } from '@nuxt/vue-app';
import { Store } from 'vuex';

export default <T = any, K = any>(store?: Store<T, K>) => {
    const Mock = jest.fn<Context, any>().mockImplementation(() => ({
        store,
    } as any));
    return new Mock();
};
