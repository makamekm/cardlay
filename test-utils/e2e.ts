import { toMatchImageSnapshot } from 'jest-image-snapshot';
import 'jest-environment-puppeteer';

jest.setTimeout(60000);
expect.extend({ toMatchImageSnapshot });

export const find = async (selector) => {
  await page.waitForSelector(selector);
  return page.$(selector);
};

export const findAll = async (selector) => {
  await page.waitForSelector(selector);
  return page.$$(selector);
};

export const containsText = async (text, wrapperSelector) => {
  const textSelector = `//text()[contains(.,'${text}')]`;
  const wrapper = wrapperSelector ? await find(wrapperSelector) : page;
  await page.waitForXPath(textSelector);
  const matches = await wrapper!.$x(textSelector);
  return matches.length > 0;
};

export const open = async url => {
  await page.setViewport({ width: 1024, height: 768 });
  await page.goto(`http://${process.env.HOST}:${process.env.PORT}${url}`, { waitUntil: `domcontentloaded` });
  await page.waitFor(1000);
};
