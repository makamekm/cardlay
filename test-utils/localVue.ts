import { createLocalVue } from '@vue/test-utils';
import VeeValidate from 'vee-validate';
import Vuex from 'vuex';
import Buefy from 'buefy';

export default () => {
    const vue = createLocalVue();
    vue.use(VeeValidate);
    vue.use(Vuex);
    vue.use(Buefy);
    return vue;
};
