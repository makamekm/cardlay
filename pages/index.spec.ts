import { shallowMount, mount } from '@vue/test-utils';
import page from './index.vue';
import getLocalVue from '~/test-utils/localVue';
import getContext from '~/test-utils/context';
import { Store } from 'vuex';
import VueRouter from 'vue-router';

describe('page index', () => {

  const getData = () => ([{
    id: 'fc7f8213',
    date: '2019-08-12T09:18:30Z',
    amount: 1324.00,
    merchant: 'vitales',
    currency: 'DKK',
    categoryName: 'Food',
    parentId: null,
    status: {
      stage: 'Submitted',
    },
    approved: true,
  }]);

  const getStore = (nodes: any[] = []) => {
    const onLoad = jest.fn();
    return {
      onLoad,
      store: new Store({
        modules: {
          LocalStorage: {
            namespaced: true,
            state: () => ({
              expenses: [],
            }),
            actions: {
              setExpenses: jest.fn(),
            },
          },
          expenses: {
            namespaced: true,
            state: () => ({
              nodes,
              limit: 10,
              page: 1,
              sortBy: null,
              sortByDescending: false,
            }),
            actions: {
              loadItems: onLoad,
              setPage: jest.fn(),
              toggleSortBy: jest.fn(),
            },
            getters: {
              sortedItems: () => [],
              total: () => 0,
            },
          },
        },
      }),
    };
  };

  const getRouter = (onPushCall) => {
    const onPush = jest.fn((path) => {
      onPushCall(path);
    });
    const RouterMock = jest.fn<VueRouter, any>().mockImplementation(() => ({
      push: onPush,
    } as any));
    return {
      onPush,
      router: new RouterMock(),
    };
  };

  test('should mounts properly', () => {
    const localVue = getLocalVue();
    const { store } = getStore();
    const wrapper = shallowMount(page, {
      localVue,
      store,
    });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test('should renders properly with no data', () => {
    const localVue = getLocalVue();
    const { store } = getStore();
    const wrapper = mount(page, {
      localVue,
      store,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should renders properly with data', () => {
    const localVue = getLocalVue();
    const { store } = getStore(getData());
    const wrapper = mount(page, {
      localVue,
      store,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should call fetch', async () => {
    const localVue = getLocalVue();
    const { store, onLoad } = getStore();
    const wrapper = shallowMount(page, {
      localVue,
      store,
    });
    await wrapper.vm.$options.fetch!(getContext(store));
    expect(onLoad).toBeCalled();
  });

  // test('should open page node/1', async () => {
  //   const localVue = getLocalVue();
  //   const { store } = getStore(getData());
  //   let route;
  //   const { router, onPush } = getRouter(r => route = r);
  //   const wrapper = mount(page, {
  //     localVue,
  //     store,
  //     mocks: {
  //       $router: router,
  //     },
  //   });
  //   await wrapper.vm.$options.fetch!(getContext(store));
  //   const node = wrapper.find('tbody tr.lc-pointer');
  //   node.trigger('click');
  //   expect(onPush).toBeCalled();
  //   expect(route.path).toBe('/node/1');
  // });
});
