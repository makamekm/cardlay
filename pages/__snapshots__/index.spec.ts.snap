// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page index should renders properly with data 1`] = `
<div>
  <div class="hero is-primary p-b-6 p-t-6 p-l-7 p-r-7 on-appear-slide-down on-delay-0">
    <div class="has-text-weight-bold is-size-5" style="letter-spacing: 1px;">
      Expenses
    </div>
  </div>
  <div class="table-container on-appear-slide-up on-delay-1">
    <table class="table is-striped is-fullwidth">
      <thead>
        <tr class="on-appear-scale-left on-delay-3">
          <th>Date</th>
          <th class="interactable">
            <div class="columns is-vcentered">
              <div class="column">
                Merchant
              </div>
              <div class="column is-narrow font-size-0"><i class="i is-size-6 sort"></i></div>
            </div>
          </th>
          <th class="interactable">
            <div class="columns is-vcentered">
              <div class="column">
                Category
              </div>
              <div class="column is-narrow font-size-0"><i class="i is-size-6 sort"></i></div>
            </div>
          </th>
          <th class="interactable">
            <div class="columns is-vcentered is-variable is-3">
              <div class="column has-text-right">
                Amount
              </div>
              <div class="column is-narrow font-size-0"><i class="i is-size-6 sort"></i></div>
            </div>
          </th>
          <th class="has-text-right">Valuta</th>
          <th>Status</th>
        </tr>
      </thead>
      <tbody name="table-row-on" duration="[object Object]"></tbody>
      <tfoot>
        <tr>
          <td colspan="6" class="p-t-8 p-b-8 has-text-centered">
            <div class="columns is-centered on-appear-slide-up on-delay-4">
              <div class="column is-one-third">
                <nav class="pagination"><a role="button" href="#" disabled="disabled" aria-label="Previous page" class="pagination-previous"><span class="icon" aria-hidden="true"><i class="fa fa-angle-left fa-lg"></i></span></a> <a role="button" href="#" disabled="disabled" aria-label="Next page" class="pagination-next"><span class="icon" aria-hidden="true"><i class="fa fa-angle-right fa-lg"></i></span></a>
                  <ul class="pagination-list">
                    <!---->
                    <!---->
                    <!---->
                    <!---->
                  </ul>
                  <!---->
                </nav>
              </div>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;

exports[`page index should renders properly with no data 1`] = `
<div>
  <div class="hero is-primary p-b-6 p-t-6 p-l-7 p-r-7 on-appear-slide-down on-delay-0">
    <div class="has-text-weight-bold is-size-5" style="letter-spacing: 1px;">
      Expenses
    </div>
  </div>
  <div class="table-container on-appear-slide-up on-delay-1">
    <table class="table is-striped is-fullwidth">
      <thead>
        <tr class="on-appear-scale-left on-delay-3">
          <th>Date</th>
          <th class="interactable">
            <div class="columns is-vcentered">
              <div class="column">
                Merchant
              </div>
              <div class="column is-narrow font-size-0"><i class="i is-size-6 sort"></i></div>
            </div>
          </th>
          <th class="interactable">
            <div class="columns is-vcentered">
              <div class="column">
                Category
              </div>
              <div class="column is-narrow font-size-0"><i class="i is-size-6 sort"></i></div>
            </div>
          </th>
          <th class="interactable">
            <div class="columns is-vcentered is-variable is-3">
              <div class="column has-text-right">
                Amount
              </div>
              <div class="column is-narrow font-size-0"><i class="i is-size-6 sort"></i></div>
            </div>
          </th>
          <th class="has-text-right">Valuta</th>
          <th>Status</th>
        </tr>
      </thead>
      <tbody name="table-row-on" duration="[object Object]"></tbody>
      <tfoot>
        <tr>
          <td colspan="6" class="p-t-8 p-b-8 has-text-centered">
            <div class="columns is-centered on-appear-slide-up on-delay-4">
              <div class="column is-one-third">
                <nav class="pagination"><a role="button" href="#" disabled="disabled" aria-label="Previous page" class="pagination-previous"><span class="icon" aria-hidden="true"><i class="fa fa-angle-left fa-lg"></i></span></a> <a role="button" href="#" disabled="disabled" aria-label="Next page" class="pagination-next"><span class="icon" aria-hidden="true"><i class="fa fa-angle-right fa-lg"></i></span></a>
                  <ul class="pagination-list">
                    <!---->
                    <!---->
                    <!---->
                    <!---->
                  </ul>
                  <!---->
                </nav>
              </div>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
